File: algorithms/cracking_the_coding_interview/16_moderate/EnglishInt.hs
import Data.List (intercalate)
import Control.Monad (forM_)

smalls = [ ""
         , "One"
         , "Two"
         , "Three"
         , "Four"
         , "Five"
         , "Six"
         , "Seven"
         , "Eight"
         , "Nine" 
         , "Ten"
         , "Eleven"
         , "Twelve"
         , "Thirteen"
         , "Fourteen"
         , "Fifteen"
         , "Sixteen"
         , "Seventeen"
         , "Eighteen"
         , "Nineteen" ]
tens = [ ""
       , ""
       , "Twenty"
       , "Thirty"
       , "Forty"
       , "Fifty"
       , "Sixty"
       , "Seventy"
       , "Eighty"
       , "Ninety" ]

hundred = "Hundred"

units = [ ""
        , "Thousand"
        , "Million"
        , "Billion" ]

toEnglish :: Int -> String
toEnglish n 
    | n < 0 = toString $ ["Negative", toEnglish (-n)]
    | n == 0 = "Zero"
    | n <= 19 = smalls !! n
    | n < 100 = toString $ if n `mod` 10 /= 0 
                           then [tens !! (n `div` 10), toEnglish (n `mod` 10)]
                           else [tens !! (n `div` 10)]
    | n < 1000 = toString $ 
                 if n `mod` 100 /= 0
                 then [ toEnglish (n `div` 100)
                      , hundred
                      , toEnglish (n `mod` 100)]
                 else [toEnglish (n `div` 100), hundred]
    | otherwise = toString $ reverse $ 
                  zipWith concatStr (map toEnglish $ toThreeDigit n) units
    where
        toString = intercalate " "
        toThreeDigit :: Int -> [Int]
        toThreeDigit x
            | x < 1000 = [x]
            | otherwise = let rem = x `mod` 1000
                              x' = x `div` 1000
                          in if x' == 0
                             then [rem] 
                             else rem : (toThreeDigit x')
        concatStr a b
            | null b = a
            | otherwise = toString [a, b]

main = do
    forM_ [1..99] $ \i -> print (i, toEnglish i)
    forM_ [123, 1234, 12345, 123456, 1234567, 12345678, 123456789] $ \i ->
        print (i, toEnglish i)

File: algorithms/cracking_the_coding_interview/16_moderate/english_int.py
# given any integer, print an English phrase that describes the integer
# e.g. One Thousand, Two Hundred Thirty-Four

below_20 = ['Zero', 'One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven',
            'Eight', 'Nine', 'Ten', 'Eleven', 'Twelve', 'Thirteen',
            'Fourteen', 'Fifteen', 'Sixteen', 'Seventeen', 'Eighteen',
            'Nineteen']
tens = [None, None, 'Twenty', 'Thirty', 'Fourty', 'Fifty', 'Sixty', 'Seventy',
        'Eighty', 'Ninety']

units = [None, 'Thousand', 'Million', 'Billion']

import random

def int_to_english(n):
    if n < 20: return below_20[n]
    if n < 100:
        return (tens[n//10] if n%10 == 0
                else tens[n//10] + '-' + below_20[n%10])
    if n < 1000:
        return (below_20[n//100] + ' Hundred' if n%100 == 0
                else below_20[n//100] + ' Hundred ' + int_to_english(n%100))

    # anything that's greater then 1000
    assert n < 1000000000000 # less than 1000 billion for testing purpose
    stack = []
    while n%1000 > 0:
        tail = n%1000
        n = n//1000
        stack.append(int_to_english(tail))

    #print(stack)
    assert len(stack) <= 3
    ret = []
    for i in reversed(range(1, len(stack))):
        #print(i)
        ret.append(stack[i] + ' ' + units[i])
    ret.append(stack[0])
    return ', '.join(ret)
        

if __name__ == "__main__":
    for i in range(1000):
        print(i, int_to_english(i))
                

    for i in (1234, 12345, 123456, 1234567, 12345678, 123456789):
        print(i, int_to_english(i))
